Show time remaining estimate during conversion

allewun opened the pull request:
Fixes #52

<img src="https://user-images.githubusercontent.com/1256911/54660900-711fc200-4a95-11e9-9942-aa773489655d.gif" width="400" />

sindresorhus commented:
Thanks for working on this 👍

sindresorhus commented:
Note this from the original issue, which should be handled:

> Yeah, there will be some uneven progress in the beginning (roughly for the first N frames where N = cpu cores), but then it should be smooth.

sindresorhus commented on Gifski/MainWindowController.swift:
```diff
@@ -130,6 +141,10 @@ final class MainWindowController: NSWindowController {
 
+	private var progressObserver: NSKeyValueObservation?
+	private var indeterminateObserver: NSKeyValueObservation?
+	private var startTime = NSDate()
```
Why not use `Date()`?

allewun replied:
muscle memory :)

sindresorhus commented:
I don't think we should show `Calculating time remaining…`. Would be better to just wait to fade in the time remaining label when we have a guesstimate.

sindresorhus commented:
I think we should only show seconds when less than 1 minutes.

allewun commented:
<img src="https://user-images.githubusercontent.com/1256911/54665922-ac29f180-4aa5-11e9-9ebb-28115d8cbc17.gif" width="400" />

allewun commented:
> Note this from the original issue, which should be handled:
> 
> > Yeah, there will be some uneven progress in the beginning (roughly for the first N frames where N = cpu cores), but then it should be smooth.

Regarding this, it seems unavoidable that the estimation will be inaccurate at start. Browser file downloads, torrent downloads, video encoding etc.. all tend to behave similarly from what I've seen. We could try waiting a few more seconds before showing the progress so it feels more accurate overall.

sindresorhus commented:
> We could try waiting a few more seconds before showing the progress so it feels more accurate overall.

Yeah, I think that would be good.

sindresorhus commented:
And after waiting a few more seconds, if the estimated progress is 10 seconds or less, I don't think we should show the time estimate at all.

sindresorhus commented:
I think we should fade it out when the time estimate has reached 1 seconds or less, as the below doesn't look very good (It stays there until the progress circle also fades out):

<img width="472" alt="Screen Shot 2019-03-20 at 16 13 12" src="https://user-images.githubusercontent.com/170270/54672817-952de280-4b2b-11e9-8a10-280558bf5530.png">

sindresorhus commented on Gifski/MainWindowController.swift:
```diff
@@ -186,6 +217,30 @@ final class MainWindowController: NSWindowController {
+	}
+
+	private func formattedTimeRemaining(seconds: TimeInterval) -> String? {
+		elapsedTimeFormatter.allowedUnits = seconds < 60 ? .second : .minute
```
It should still show hours.

```suggestion
		elapsedTimeFormatter.allowedUnits = seconds < 60 ? .second : [.hour, .minute]
```

allewun commented:
@sindresorhus I moved the estimation logic into its own class and implemented:
1. buffer for 3 seconds before showing estimate
2. if initial estimate is less than 10 seconds, don't bother showing it at all
3. fade the estimate 1 second before finishing


**Long processing, show label after estimate stabilizes**:
![buffer](https://user-images.githubusercontent.com/1256911/54862267-56e41f00-4cf5-11e9-8f6f-935023e4ef44.gif)

**Short processing, don't show estimate**:
![skip](https://user-images.githubusercontent.com/1256911/54862268-56e41f00-4cf5-11e9-9e49-d333a3d3cbc3.gif)

sindresorhus commented:
This is looking great now. Thanks for working on this, @allewun 🙌

allewun commented:
@sindresorhus thanks for working with me on it!

sindresorhus commented:
It will be out very soon. It’s waiting for review.

geordiemhall commented:
Great change! 

Very minor thing - might be nicer to round away from zero for the time shown in the label so that it never hits "0 seconds remaining" (would just remain on "1 second remaining" during fade out).

sindresorhus commented:
@geordiemhall I agree, and already did that in a follow-up commit: https://github.com/sindresorhus/gifski-app/commit/fc5482ee9c74f26b5281f4c80d848626d8c3dc1e#diff-81f725e4fa6f9c266d8d01bc0c462570R104

geordiemhall commented:
Ah awesome, didn’t notice sorry!
